kuma/wiki/forms.py

```diff
@@ -233,47 +250,55 @@ def __init__(self, *args, **kwargs):
         if self.section_id:
             self.fields['toc_depth'].required = False
 
-    def _clean_collidable(self, name):
-        value = self.cleaned_data[name]
-
+    def clean_slug(self):
+        # Since this form can change the URL of the page on which the editing
+        # happens, changes to the slug are ignored for an iframe submissions
         if self.is_iframe_target:
-            # Since these collidables can change the URL of the page, changes
-            # to them are ignored for an iframe submission
-            return getattr(self.instance.document, name)
+            return self.instance.document.slug
+
+        # Get the cleaned slug
+        slug = self.cleaned_data['slug']
+
+        # first check if the given slug doesn't contain slashes and other
+        # characters not allowed in a revision slug component (without parent)
+        if slug and INVALID_REV_SLUG_CHARS_RE.search(slug):
+            raise forms.ValidationError(SLUG_INVALID)
+
+        # edits can come in without a slug, so default to the current doc slug
+        if not slug:
+            try:
+                slug = self.instance.slug = self.instance.document.slug
+            except ObjectDoesNotExist:
+                pass
+
+        # then if there is a parent document we prefix the slug with its slug
+        if self.parent_slug:
+            slug = u'/'.join([self.parent_slug, slug])
 
-        error_message = {'slug': SLUG_COLLIDES}.get(name, OTHER_COLLIDES)
         try:
-            existing_doc = Document.objects.get(
-                locale=self.instance.document.locale,
-                **{name: value})
+            doc = Document.objects.get(locale=self.instance.document.locale,
```

Comment:
I'm trying to wrap my brain around this try/except block, so bear with me.

- No existing ``Document`` in locale with the slug - ``Document.DoesNotExist`` is raised, slug is OK to use
- There is an existing ``Document`` with slug...
  - and ``Revison`` has a ``Document``...
    - and the existing ``Document`` is a redirect: slug was used in the past, but is OK to reuse
    - and the existing ``Document`` is not a redirect...
       - and does not match ``Revision``'s ``Document``: ``ValidationError(SLUG_COLLIDES)`` is raised
       - and matches ``Revision``'s ``Document``: slug is OK, colliding document is same document
  - and ``Revision`` doesn't exist yet: ``ValidationError(SLUG_COLLIDES)`` is raised
  - and ``Revision`` doesn't have a document yet: ``ValidationError(SLUG_COLLIDES)`` is raised
  - and accessing ``Revision``'s document raises a ``Document.DoesNotExist`` exception: slug is OK :interrobang:

It feels like the last case is a (pre-existing) bug, if ``self.instance.document`` can raise a ``Document.DoesNotExist`` error.  Which maybe it doesn't.  If it is a potential issue, then we'd need something like:

```python
        # Is the slug in use?
        try:
            doc = Document.objects.get(locale=self.instance.document.locale,
                                       slug=slug)
        except Document.DoesNotExist:
            # No existing document for this value, so we're good here.
            pass
        else:
            if not doc.get_redirect_url():
                rev_doc = None
                try:
                    rev_doc = self.instance and self.instance.document
                except Document.DoesNotExist:
                    pass

                if rev_doc is None:
                    # This document-and-revision doesn't exist yet, so there
                    # shouldn't be any collisions at all.
                    raise forms.ValidationError(SLUG_COLLIDES)
                elif doc.pk != rev_doc.pk:
                    # There's another document with this value,
                    # and we're not a revision of it.
                    raise forms.ValidationError(SLUG_COLLIDES)
```

Even if ``self.instance.document`` can't raise an exception, I like my except blocks as tight as possible.